### PR TITLE
feat(2196): secrets management 

### PR DESCRIPTION
**Pull request by RemingtonReackhof**
Implemented #2196 to add support for tiller to use secrets to store releases. This includes support for `secrets.yaml` use alongside of `values.yaml`. 

Reference these secrets in templates the same as values, such as `{{ .Secrets.type }}`

Also adds support for arguments to `install` and `upgrade` identical to those used by values, e.g.: 
`--set-secrets type=ClusterIP` and 
`--secrets=mysecrets.yaml`

For use of secrets storage backend in Tiller, a new option was added to the `--storage` argument when starting Tiller:

`tiller --storage=secret`

### CODE REVIEW

**Comment by thomastaylor312** [comment]
@RemingtonReackhof Thank you for submitting this! This is a pretty big feature add, so several of us are going to have to review this before merge.

**Comment by thomastaylor312** [comment]
@RemingtonReackhof Just to update you on this. We will be reviewing this and merging in once we have released 2.6. Most of our efforts are focused on 2.6 right now

**Comment by RemingtonReackhof** [comment]
@thomastaylor312 I appreciate the update, best of luck with the 2.6 release!

**Comment by thomastaylor312** [comment]
@RemingtonReackhof I plan on getting to this no later than Monday, but hopefully tomorrow. I want to make sure I put adequate time into reviewing this well

**Comment by thomastaylor312** [comment]
Ok, I've started on this. It is going to take a bit to go all through

File: _proto/hapi/services/tiller.proto
Comment (thomastaylor312) [inline _proto/hapi/services/tiller.proto]: Nit: Can this comment be a bit more descriptive?

File: cmd/helm/install_test.go
Comment (thomastaylor312) [inline cmd/helm/install_test.go]: I assume you meant this to be `strings.Split("--set-secrets foo=bar --set-secrets bar=foo", " ")`?

File: cmd/helm/upgrade.go
Comment (thomastaylor312) [inline cmd/helm/upgrade.go]: Make sure to pass `secretFiles` and `secrets` to the `installCmd` above. Otherwise if a user does `helm upgrade --install`, they won't end up using the secrets

File: pkg/chartutil/load.go
Comment (thomastaylor312) [inline pkg/chartutil/load.go]: Do you also need to set `c.Values.Type` here?

File: pkg/chartutil/save.go
Comment (thomastaylor312) [inline pkg/chartutil/save.go]: Because these are secrets, it may be better to do `500` for the permissions

File: pkg/chartutil/values.go
Comment (thomastaylor312) [inline pkg/chartutil/values.go]: Can you update this comment to reflect the secrets change?

File: pkg/chartutil/values.go
Comment (thomastaylor312) [inline pkg/chartutil/values.go]: This is a breaking public API change. This doesn't mean we can't pull it in to 2.X, but we have to have a consensus among all of us that this is ok.

File: pkg/storage/driver/mock_test.go
Comment (thomastaylor312) [inline pkg/storage/driver/mock_test.go]: Nit: Comment looks like it has a copy/paste error

File: pkg/storage/driver/mock_test.go
Comment (thomastaylor312) [inline pkg/storage/driver/mock_test.go]: Should be "Failed to create secret"

File: pkg/storage/driver/secrets.go
Comment (thomastaylor312) [inline pkg/storage/driver/secrets.go]: Because these are used in Secrets and ConfigMaps, it may be useful to define them as constants

File: pkg/storage/driver/secrets.go
Comment (thomastaylor312) [inline pkg/storage/driver/secrets.go]: Should be 2017 in this and a couple other of the new files 🙂

File: pkg/storage/driver/secrets.go
Comment (thomastaylor312) [inline pkg/storage/driver/secrets.go]: This and the next var look like copy/paste remnants.

**Review by thomastaylor312 — COMMENTED** [review]
I have not tested this yet (will do soon), but wanted to get you my feedback on the code. Mostly small comments or nits.

I am going to label this as breaking because of one of the changes could be considered such.

**Comment by RemingtonReackhof** [comment]
@thomastaylor312 Thank you for the comments! I apologize for the late response I've been without a computer for the last week. I'll get on these changes asap

File: pkg/chartutil/load.go
Comment (RemingtonReackhof) [reply pkg/chartutil/load.go]: The type will default to `chart.ConfType_VALUES` since its initialized to `0` but I can make it explicit if that would be clearer?

File: pkg/storage/driver/mock_test.go
Comment (RemingtonReackhof) [reply pkg/storage/driver/mock_test.go]: Looks like line 108 of the same file has the same typo. Not entirely sure what the word should be, probably `list` ?

**Comment by bacongobbler** [comment]
once conflicts/comments have been addressed, I'd be happy to take a look at this as well.

File: cmd/helm/install.go
Comment (bacongobbler) [inline cmd/helm/install.go]: is there any way we can make it explicit that these flags only work with tiller set to --storage=secret?

**Review by bacongobbler — COMMENTED** [review]
I'm concerned about the design RE: the CLI flags. Why can't we just continue to use --set and --values when the storage type is set to secrets?

I'm also concerned that this will make other storage layers like `inmemory` (the testing driver) and configmap (the stable storage backend) degenerate use cases since they can no longer fulfill all of the feature flags the secret storage layer provides. Perhaps we should consider tabling this until 3.0.

**Comment by thomastaylor312** [comment]
Now that @bacongobbler mentions it, I think that using `--set` and `--values` should be sufficient here. There is nothing different about a `secrets.yaml` and a `values.yaml` except for the fact that you shouldn't commit secrets. Thoughts @RemingtonReackhof?

File: cmd/tiller/tiller.go
Comment (thomastaylor312) [inline cmd/tiller/tiller.go]: The `store` flag needs to be updated to show that `secret` is now an option

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
One tiny little change I discovered during testing. Otherwise testing all checks out

**Comment by RemingtonReackhof** [comment]
@thomastaylor312 @bacongobbler I agree with those points, the added CLI flags aren't necessary. I think the most important part of this PR is adding the secrets storage for releases. I can go ahead and remove those changes if that's what we decide is best.

**Comment by thomastaylor312** [comment]
@RemingtonReackhof I am thinking we remove those flags for now. That ok?

File: pkg/storage/driver/mock_test.go
Comment (thomastaylor312) [reply pkg/storage/driver/mock_test.go]: Yeah I think so. Some thing like "returns a list of secrets"

**Review by thomastaylor312 — APPROVED** [review]
I did one more round of testing as a sanity check. I tested installs, deletes, and upgrades. I also made sure I could see the secrets in k8s and all of the `helm get` features. I think this is good to go! I am going to wait for final sign off from @bacongobbler and then we can merge this in.

Great work @RemingtonReackhof!

**Comment by bacongobbler** [comment]
Everything looks good here, but the only thing missing would be some documentation. Would you mind adding documentation on using the secrets backend at https://github.com/kubernetes/helm/tree/master/docs?

**Comment by RemingtonReackhof** [comment]
I'll get on it this weekend @bacongobbler, sorry for the delay.

**Comment by thomastaylor312** [comment]
@RemingtonReackhof Any update on this? We will be cutting a RC soon and would love to see this land.

**Comment by RemingtonReackhof** [comment]
@thomastaylor312 @bacongobbler Installing tiller with a secrets storage backend requires `tiller --storage=secret` to be run; however this has some implications regarding the deployment of the tiller image into a cluster. 

Firstly, the tiller image that gets built runs with `/tiller` command. I looked into adding an option to `helm init` to use the `--storage` option for tiller, however I believe that it's only possible to change the storage argument by changing the `/tiller` command from the docker image.

I looked into this more and went ahead and changed the Dockerfile to run `CMD ["/tiller" , "--storage=secret"]` and was able to get the image to run with a secrets backend on a local cluster. From this, I believe the best option would be to build a separate tiller image that uses the secrets backend command and then have users run `helm init --tiller-image={{tiller-secrets-backend-image}}`. 

I realize that this solution is not ideal but I'm at a loss as to another way. If you have any insight, that would be much appreciated.

**Comment by thomastaylor312** [comment]
@RemingtonReackhof It isn't too much to ask someone to edit a line in the tiller deployment, but you are right in saying that it isn't ideal. We'll be adding some other flags to helm init that allow you to customize it a bit, but for now, I would just have people edit the deployment manually. 

@bacongobbler thoughts?

**Comment by jakehow** [comment]
@RemingtonReackhof  if you added an option to `helm init` you could modify the `generateDeployment` function in [install.go](https://github.com/kubernetes/helm/blob/a736c2a3fe9a0f452cb18038a301f5fe9fb026c8/cmd/helm/installer/install.go#L120)

I think you could just have it set the command that is used with the image explicitly in the deployment.yaml if the flag is present.   Separate images shouldn't be necessary.

**Comment by thomastaylor312** [comment]
@RemingtonReackhof So we will be merging #2557 soon. That has support for better manipulation of the Tiller manifest. You can see the docs [here](https://github.com/kubernetes/helm/pull/2557/files#diff-64407d3e6d5ef382426bdfbcae040c6aR200) and can use that as a basis. Probably something using `--output` or `--set`

**Comment by thomastaylor312** [comment]
@RemingtonReackhof #2557 was merged and the docs are now available: https://github.com/kubernetes/helm/blob/master/docs/install.md#advanced-usage

**Comment by jakehow** [comment]
@thomastaylor312 with that PR it seems like we should be able to document the option and just do:

`helm init --override "spec.template.spec.containers[0].command"="tiller --storage=secret"`

or if using curly braces works like it does for `--set`:

`helm init --override "spec.template.spec.containers[0].command"={"tiller", "--storage=secret"}`

**Comment by redbaron** [comment]
Forgive me if I miss anything, but it looks like it doesn't implement original proposal where secrets are stored as hashes and are not retrievable by subseqent calls.  If it is the case, then only difference between this storage driver and ConfigMap stroage driver is `.impl` type. Can't we alter ConfigMaps to support storing data in Secrets if needed?

There is also no conversion from old storage to new, so I can see how it will cause major pain for anybody willing to enable this feature in existing setups.

**Comment by thomastaylor312** [comment]
@jakehow That is correct, should be pretty simple.

@redbaron This is just the first step of that proposal. Now that encrypted secrets are an option, some people will want to use it and we can implement the rest of the proposal later on. As for the conversion, that is a good point, but I am not sure if that is something that needs to go in to the release itself, but it may be useful to create a script to use out of band. Thoughts?

**Comment by bacongobbler** [comment]
> There is also no conversion from old storage to new, so I can see how it will cause major pain for anybody willing to enable this feature in existing setups.

> As for the conversion, that is a good point, but I am not sure if that is something that needs to go in to the release itself, but it may be useful to create a script to use out of band.

I think it's completely fine (and I would prefer) to ship this as a new "experimental" storage backend with no migration story. This is already a massive addition to the codebase (thank you @RemingtonReackhof, btw) and will likely need some battle testing before we go ahead and start writing data migration scripts from configmaps to secrets. If anything that should be a follow-up PR or an out-of-band script like @thomastaylor312 mentions IMO. If we're writing data migration scripts, it's probably best to hold that off until Helm 3.

**Comment by thomastaylor312** [comment]
@RemingtonReackhof We discussed this in the dev call and decided we are going to go ahead and merge. You have already spent a good deal of time on this PR, so one of us can add the documentation in a separate PR.

**Comment by redbaron** [comment]
@bacongobbler , @thomastaylor312  this "massive addition" is a copypaste of a configmap driver, there is simply nothing to battle test, it is as robust as old driver. It adds very little value without a migration path :(

```
$ sed -i -e 's,secrets,cfgmaps,g' -e 's,Secret,ConfigMap,g' -e 's,//.*,,g' secrets.go    # rename IDs, delete comments 
$ sed -i -e 's,//.*,,g' cfgmaps.go   # delete comments
$ diff -u cfgmaps.go secrets.go
--- cfgmaps.go  2017-10-13 07:35:06.714096200 +0100
+++ secrets.go  2017-10-13 07:34:55.949777000 +0100
@@ -1,5 +1,5 @@
 /*
-Copyright 2016 The Kubernetes Authors All rights reserved.
+Copyright 2017 The Kubernetes Authors All rights reserved.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
@@ -72,7 +72,7 @@
                return nil, err
        }

-       r, err := decodeRelease(obj.Data["release"])
+       r, err := decodeRelease(string(obj.Data["release"]))
        if err != nil {
                cfgmaps.Log("get: failed to decode data %q: %s", key, err)
                return nil, err
@@ -99,7 +99,7 @@


        for _, item := range list.Items {
-               rls, err := decodeRelease(item.Data["release"])
+               rls, err := decodeRelease(string(item.Data["release"]))
                if err != nil {
                        cfgmaps.Log("list: failed to decode release: %v: %s", item, err)
                        continue
@@ -136,7 +136,7 @@

        var results []*rspb.Release
        for _, item := range list.Items {
-               rls, err := decodeRelease(item.Data["release"])
+               rls, err := decodeRelease(string(item.Data["release"]))
                if err != nil {
                        cfgmaps.Log("query: failed to decode release: %s", err)
                        continue
@@ -164,7 +164,7 @@

        if _, err := cfgmaps.impl.Create(obj); err != nil {
                if apierrors.IsAlreadyExists(err) {
-                       return ErrReleaseExists(key)
+                       return ErrReleaseExists(rls.Name)
                }

                cfgmaps.Log("create: failed to create: %s", err)

**Comment by thomastaylor312** [comment]
@redbaron I have to disagree on that point (although it is very similar code to the config map backend). It definitely is a value add mainly because you can now encrypt `Secret` objects by default and there is the possibility of future integration of secrets with a backend like Vault. See the discussion starting [here](https://github.com/kubernetes/helm/issues/2196#issuecomment-317557164) for reference.

**Comment by marty-stranger** [comment]
`<.Secrets.global.aws....>: can't evaluate field global in type interface {}`
global is not supported for Secrets?

**Comment by armab** [comment]
@marty-stranger According to code, the PR only implements `tiller --storage=secret` and there is a corresponding documentation change for that: https://github.com/kubernetes/helm/pull/3034

I couldn't find (as advertised in PR description) implementation for `secrets.yaml`, `{{ .Secrets.type }}` and `--set-secrets`/`--secrets` options for `helm install` and `helm upgrade`.

^^ That's a confusing part about this PR to know about.

**Comment by bitsofinfo** [comment]
How do we reference the `{{ .Secrets....}}`? or is that still in progress

**Comment by Magizhchi** [comment]
> @marty-stranger According to code, the PR only implements `tiller --storage=secret` and there is a corresponding documentation change for that: #3034
> 
> I couldn't find (as advertised in PR description) implementation for `secrets.yaml`, `{{ .Secrets.type }}` and `--set-secrets`/`--secrets` options for `helm install` and `helm upgrade`.
> 
> ^^ That's a confusing part about this PR to know about.

Is there any guidance on this, I am also looking to use `{{ .Secrets.mySecretValue }}` in my charts

**Pull request merged**
